lib/drivers/docker/index.js

```diff
@@ -0,0 +1,294 @@
+var EventEmitter = require('events').EventEmitter;
+var Image = require('./image');
+var inherits = require('util').inherits;
+var isError = require('util').isError;
+
+module.exports = exports = DockerDriver;
+
+function DockerDriver(opts) {
+  if (!opts.baseDir || !opts.console || !opts.server) {
+    throw Error('Requires baseDir, console, and server');
+  }
+  if (!(this instanceof DockerDriver)) {
+    return new DockerDriver(opts);
+  }
+  EventEmitter.call(this);
+  this.docker = new Docker();
+
+  this.baseDir = opts.baseDir;
+  this.console = opts.console;
+  this.server = opts.server;
+
+  this.instances = {};
+
+  this.defaultStartOptions = {
+    profile: true,
+    trace: false,
+    size: 'STRONGLOOP_CLUSTER' in process.env ?
+      process.env.STRONGLOOP_CLUSTER : 'CPU',
+  };
+
+  this.on('request', this._onRequest.bind(this));
+}
+
+inherits(DockerDriver, EventEmitter);
+
+DockerDriver.prototype.setStartOptions = function(id, opts) {
+  debug('DockerDriver.setStartOptions(%j, %j)', id, opts);
+  // console.trace();
+  var instance = this._instance(id);
+  var original = _.clone(instance.startOpts);
+
+  instance.startOpts = _.merge(instance.startOpts, opts);
+
+  if (opts.size != null && opts.size !== original.size) {
+    debug('cluster size changed, sending set-size: %j => %j', id, opts);
+    this.requestOfInstance(id, {cmd: 'set-size', size: opts.size});
+  } else {
+    // debug('cluster size unchanged');
+  }
+};
+
+DockerDriver.prototype.removeInstance = function(id, cb) {
+  debug('DockerDriver.removeInstance(%j)', id);
+  var instance = this._instance(id);
+  if (instance.current) {
+    instance.current.kill('SIGKILL', cb);
+    delete instance.current;
+  } else {
+    setImmediate(cb);
+  }
+};
+
+DockerDriver.prototype.deployInstance = function(id, req, res) {
+  debug('DockerDriver.deployInstance(%j)', id);
+  var self = this;
+  var instance = this._containerById(id);
+  var image = Image.from(this, id, this.baseDir, req, res);
+  instance.next = image;
+  image.on('error', function(err) {
+    console.error('error deploying: %s, %j', err, err);
+    throw err;
+  });
+  image.on('image', function() {
+    debug('Deployment image:', image.name);
+    self.instances[id].next.image = image;
+    self.startInstance(id, function(err) {
+      debug('deployInstance -> startInstance -> ', err);
+    });
+  });
+};
+
+DockerDriver.prototype.startInstance = function(id, cb) {
+  debug('DockerDriver.startInstance(%j, %j)', id, cb.name);
+  var instance = this._instance(id);
+  var toLaunch = instance.next || instance.currentImage || instance.previous;
+  var img = toLaunch.image;
+  var self = this;
+  this.server.getInstanceEnv(id, function(err, env) {
+    if (err) return cb(err);
+    // ensure environment variable values have been stringified
+    env = _.mapValues(env, String);
+    var opts = _.merge(instance.startOpts, {env: env});
+    instance.log.enableGC();
+    var previous = instance.current;
```

Comment:
I find this quite difficult to understand. The problem with instance being a plain old "bag of properties" instead of an object with methods is that it's behaviour and interface doesn't have a place it can be described and commented on... its merely implied by reading every line of code that interacts with it and they are spread throughout this file. This method is partly internal implementation of the Instance, and partly gluing internals of the Instance, the image, to itself, all mixed together.